hbase-client/src/main/java/org/apache/hadoop/hbase/client/HMasterAsyncRegistry.java

```diff
@@ -0,0 +1,236 @@
+    // Randomize so that not every client sends requests in the same order.
+    Collections.shuffle(masterServers);
+  }
+
+  /**
+   * Util that generates a master stub for a given ServerName.
+   */
+  private MasterService.BlockingInterface getMasterStub(ServerName server) throws IOException  {
+    return MasterService.newBlockingStub(
+        rpcClient.createBlockingRpcChannel(server, User.getCurrent(), rpcTimeout));
+  }
+
+  /**
+   * Blocking RPC to fetch the meta region locations using one of the masters from the parsed list.
+   */
+  private RegionLocations getMetaRegionLocationsHelper() throws MetaRegionsNotAvailableException {
+    List<HBaseProtos.RegionLocation> result = null;
+    for (ServerName sname: masterServers) {
+      try {
+        MasterService.BlockingInterface stub = getMasterStub(sname);
+        HBaseRpcController rpcController = RpcControllerFactory.instantiate(conf).newController();
+        GetMetaLocationsResponse resp = stub.getMetaLocations(rpcController,
+            GetMetaLocationsRequest.getDefaultInstance());
+        result = resp.getLocationsList();
+      } catch (Exception e) {
+        LOG.warn("Error fetch meta locations from master {}. Trying others.", sname, e);
+      }
+    }
+    if (result == null || result.isEmpty()) {
+      throw new MetaRegionsNotAvailableException(String.format(
+          "Meta locations not found. Probed masters: %s", conf.get(CONF_KEY, DEFAULT_HOST_PORT)));
+    }
+    List<HRegionLocation> deserializedResult = new ArrayList<>();
+    result.stream().forEach(
+        location -> deserializedResult.add(ProtobufUtil.toRegionLocation(location)));
+    return new RegionLocations(deserializedResult);
+  }
+
+  /**
+   * Picks the first master entry from 'masterHortPorts' to fetch the meta region locations.
+   */
+  @Override
+  public CompletableFuture<RegionLocations> getMetaRegionLocation() {
+    CompletableFuture<RegionLocations> result = new CompletableFuture<>();
+    CompletableFuture.runAsync(() -> {
+      try {
+        result.complete(this.getMetaRegionLocationsHelper());
+      } catch (Exception e) {
+        result.completeExceptionally(e);
+      }
+    });
+    return result;
+  }
+
+  /**
+   * Blocking RPC to get the cluster ID from the parsed master list. Returns null if no active
+   * master found.
+   */
+  private String getClusterIdHelper() throws  MasterNotRunningException {
+    // Loop through all the masters serially. We could be hitting some standby masters which cannot
+    // process this RPC, so we just skip them.
+    for (ServerName sname: masterServers) {
+      try {
+        MasterService.BlockingInterface stub = getMasterStub(sname);
+        HBaseRpcController rpcController = RpcControllerFactory.instantiate(conf).newController();
+        GetClusterIdResponse resp =
+            stub.getClusterId(rpcController, GetClusterIdRequest.getDefaultInstance());
+        return resp.getClusterId();
+      } catch (IOException e) {
```

Comment:
so, what's the point separating IOException and ServiceException here?
 can you combine them?